### ISSUE TITLE
Using the extension a large main file causes a memory leak + slow feature loss

### ISSUE THREAD

**Issue by LordVGames** [Sarrus1/sourcepawn-vscode#252]
# Basic informations

- **OS:** [Windows 10]
- **VSCode version:** Latest
- **Extension version:** Latest

# Further Information

I've been working on a large .sp file (around 2k lines), and noticed that, over time, RAM usage increases but doesn't decrease, and the extension slowly loses it's functionality.

Over time, the coloring sometimes gets messed up (coloring is a few characters ahead/behind where they should be), but the extension also loses some features. It can lose it's ability to auto-complete variable/function names from the main file. Soon after, it can lose it's ability to peek call hierarchy, it can lose it's ability to handle the "go to symbol" keybind, and eventually, it can lose it's ability to autocomplete variables/functions from includes (including the base sourcemod functions/variables!)

The errors/warnings are also usually delayed from updating for a few changes, and the only way to update them right away is to switch tabs to and from the large file, or reload the window. Reloading the window also fixes the RAM usage/lost features.

In the past with this extension, this didn't happen. This file with ~2k lines is set to be the main sp file. I have tried disabling this extension, and the RAM usage increasing doesn't occur when disabled.

# To Reproduce

1. Get a file with a large line count, around ~2k lines. Having the majority of the lines be white space won't cause this issue, or at least not as bad/fast.
2. Do some edits to it. Undo-ing/redo-ing a small paste of text works. Larger amounts of text seem to make it worse. Notice the RAM usage of the extension's process steadily increasing. If done enough, it can go over 1GB. The CPU usage also increases when this happens, around 5-10% more than usual.
3. As this happens, check if the features I mentioned above still work. They may slowly stop working in a random order.

# Error messages

I noticed this every few seconds in one log while trying to get details for this issue:

<details>

<summary>Log</summary>

>[2022-08-10 14:59:50.555] [exthost] [error] [Sarrus.sourcepawn-vscode] provider FAILED
[2022-08-10 14:59:50.555] [exthost] [error] TypeError: Cannot read properties of undefined (reading 'match')
	at P7 (c:\Users\Ian\.vscode\extensions\sarrus.sourcepawn-vscode-4.0.13\dist\spIndex.js:484:106)
	at yd.provideCompletionItems (c:\Users\Ian\.vscode\extensions\sarrus.sourcepawn-vscode-4.0.13\dist\spIndex.js:483:107)
	at te.provideCompletionItems (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:87:51894)
	at f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:87:77036
	at Z._withAdapter (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:87:67207)
	at Z.$provideCompletionItems (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:87:77011)
	at o._doInvokeHandler (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:95:13691)
	at o._invokeHandler (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:95:13373)
	at o._receiveRequest (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:95:12092)
	at o._receiveOneMessage (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:95:10845)
	at f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:95:8952
	at w.invoke (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:145)
	at b.deliver (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:2266)
	at p.fire (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:1844)
	at c.fire (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:69:19006)
	at f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:111:17079
	at w.invoke (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:145)
	at b.deliver (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:2266)
	at p.fire (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:1844)
	at c.fire (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:69:19006)
	at i._receiveMessage (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:69:23593)
	at f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:69:21127
	at w.invoke (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:145)
	at b.deliver (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:2266)
	at p.fire (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:61:1844)
	at p.acceptChunk (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:69:15837)
	at f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:69:14967
	at Socket.l (f:\Visual Studio Code\resources\app\out\vs\workbench\api\node\extensionHostProcess.js:69:26303)
	at Socket.emit (node:events:390:28)
	at addChunk (node:internal/streams/readable:315:12)
	at readableAddChunk (node:internal/streams/readable:289:9)
	at Socket.Readable.push (node:internal/streams/readable:228:10)
	at Pipe.onStreamRead (node:internal/stream_base_commons:199:23)
</details>

**Comment by Sarrus1** [comment]
Hi, thanks you for the detailed issue.

Here is what I've managed to reproduce:

# At startup
![mem-profile-1](https://user-images.githubusercontent.com/63302440/184112689-c94ef762-cfae-47e9-afa4-d012f2219ec0.PNG)
![mem-profile-12](https://user-images.githubusercontent.com/63302440/184112734-e25efa21-6372-4c09-9f8d-f40ece538fb7.PNG)

# After editing a 10k+ lines file (undo and some copy pasting)
![mem-profile-2](https://user-images.githubusercontent.com/63302440/184112732-0b3d7a75-bf19-4842-b9d8-d2e9a9dd0df5.PNG)
![mem-profile-22](https://user-images.githubusercontent.com/63302440/184112737-5d1ea3e4-f7c3-433f-a761-6aa334702efa.PNG)

I think this comes from [web-tree-sitter](https://www.npmjs.com/package/web-tree-sitter). I do not see a solution to fix this issue yet, but I think I will just switch to the Node bindings.

**Comment by Sarrus1** [comment]
@LordVGames I've tried to implement some debouncing to help with this issue (it's not a perfect fix, but it helps). Please give it a try if you have time by [downloading the pre-release version from the marketplace](https://github.com/Sarrus1/sourcepawn-vscode#how-to-install-the-beta-build-).

**Comment by Sarrus1** [comment]
It seems like the only solution I see is to switch to a Rust written Language Server.

I've done some benchmarking, and for a [10K line file](https://github.com/surftimer/SurfTimer/blob/6ac0ded7b87e8b92453cadb2f52935f415fa6bbe/addons/sourcemod/scripting/surftimer/sql.sp), the time to parse using WASM is between 270ms and 300ms, while with native Rust, it's between 100ms and 110ms.

**Comment by LordVGames** [comment]
After some days of using the extension on the pre-release version, the RAM usage hasn't been fixed, but the feature loss issue seems to be better.

Most features disappear much later, at around 2GB of RAM usage, but I noticed that being able to use "Rename Symbol" disappears around 1GB of RAM. Doing edits with the "Find" panel open seems to cause the bug much quicker, but that could be a coincidence.

**Comment by LordVGames** [comment]
After some days of using the extension on the pre-release version, the RAM usage hasn't been fixed, but the feature loss issue seems to be better.

Most features disappear much later, at around 2GB of RAM usage, but I noticed that being able to use "Rename Symbol" disappears around 1GB of RAM. Doing edits with the "Find" panel open seems to cause the bug much quicker, but that could be a coincidence.

**Comment by Sarrus1** [comment]
Thanks you for the feedback. I will start working on the Rust LSP implementation ASAP

**Comment by Sarrus1** [comment]
@LordVGames can you please try v5.0.1? It's a pre-release for now, but it should fix all memory-leak related problems, and has a much much lower memory footprint alltogether.

**Comment by LordVGames** [comment]
I've tried it some, and the memory increase is **much** slower now. The memory usage still doesn't go down after spam copy/pasting some text, but it happens so slowly that it likely won't be a problem in a normal session. The coloring/parsing of the text is also seems to be much faster now.

The only big problem I noticed with this version is that the extension isn't providing symbol info anymore. I can still CTRL + click to go to the definition of a function, but I can't see it's references nor rename it with F2.
